File: 000_C# Training/C#_7 StringOperations/Java/MainClass.java
import java.util.Scanner;
import java.util.Date; 
import java.text.SimpleDateFormat;  

class MainClass
{
    public static void main(String[] args)
    {
        CompareStrings_Convenient();
        CompareStrings_Correct();
        CompareString_CaseInsensitive();
        ConcatenateStrings();
        ConcatenateStringIteratively();
        LengthOfString();
        SplitStringBySingleChar();
        SplitStringByString();
        GetFirstFourSigns();
        GetFirstFourSigns_Safe();
        GetLastFourSigns();
        GetLastFourSigns_Safe();
        GetSignsBetween();
        CheckIfStringContainsSubstring();
        GetPositionWhereSubsubstringAppearsInString();
        GetPositionWhereSubsubstringAppearsInString_FromRight();
        RestOfStringAfterKeyWord();
        ReplaceStringWithSubstring();
        DateToStringWithFormat();
        NumberToStringWithLeadingZeros();
        CreateStringWithEnterAndCarriageReturn();

        new Scanner(System.in).nextLine();
    }

    public static void CompareStrings_Convenient()
    {
        //Attention, in Java and C++, this approach does not work
        //See CompareStrings_Correct()
    }

    public static void CompareStrings_Correct()
    {
        String myString = "abc";
        String myString2 = "aef";

        if (myString.equals(myString2))
            System.out.println("Strings are equal");
        else
            System.out.println("Strings are not equal");
    }        

    public static void CompareString_CaseInsensitive()
    {
        String myString = "AbCDeF";
        String myString2 = "abcdef";

        if (myString.toLowerCase().equals(myString2.toLowerCase()))
            System.out.println("Strings are equal");
        else
            System.out.println("Strings are not equal");

        if (myString.toUpperCase().equals(myString2.toUpperCase()))
            System.out.println("Strings are equal");
        else
            System.out.println("Strings are not equal");        

    }

    public static void ConcatenateStrings()
    {
        String myString = "Lgbs";
        String myString2 = "Mnl";
        String myString3 = "Krk";

        String myString4 = "We work at " + myString + " " + myString2 + " and " + myString3;
        
        System.out.println(myString4);
    }

    public static void ConcatenateStringIteratively()
    {
        String[] myArray = new String[] {"abc", "def", "ghi", "jkl"};
        String myFinalString = "";

        for (String singleString : myArray)
            myFinalString += singleString + "__";  

        System.out.println(myFinalString);
    }

    public static void LengthOfString()
    {
        String myString = "LgbsMnlKrk";
        int length = myString.length();
        System.out.println(length);
    }

    public static void SplitStringBySingleChar()
    {
        String myString = "abc;def;ghi;jkl";

        String[] mySplitString = myString.split(";");

        for (String singleString : mySplitString)
            System.out.println(singleString);
    }

    public static void SplitStringByString()
    {
        String myString = "abc;_def;_ghi;_jkl";

        String[] mySplitString = myString.split(";_");

        for (String singleString : mySplitString)
            System.out.println(singleString);        
    }

    public static void GetFirstFourSigns()
    {
        String myString = "LgbsKrkMnl";
        String leftFourSigns = myString.substring(0, 4); 
        //Attention: C#: Substring(index first signs, number of signs)
        //          Java: substring(index first sign, index last sign + 1)
        System.out.println(leftFourSigns);
        // Attention, when myString has less than four signs, an error will be thrown. 
        // See GetFirstFourSigns_Safe() to handle this problem
    }

    public static void GetFirstFourSigns_Safe()
    {
        String myString = "LgbsKrkMnl";
        String leftFourSigns = "";

        if (myString.length() >= 4)
            leftFourSigns = myString.substring(0, 4);
        else
            leftFourSigns = myString;

        System.out.println(leftFourSigns);
    }

    public static void GetLastFourSigns()
    {
        String myString = "LgbsKrkMnl";
        String rightFourSigns = myString.substring(myString.length() - 4);
        System.out.println(rightFourSigns);
    }

    public static void GetLastFourSigns_Safe()
    {
        String myString = "LgbsKrkMnl";
        String rightFourSigns = "";
    
        if (myString.length() >= 4)
            rightFourSigns = myString.substring(myString.length() - 4);
        else
            rightFourSigns = myString;

        System.out.println(rightFourSigns);
    }

    public static void GetSignsBetween()
    {
        String myString = "LgbsKrkMnl";
        String thirdToFifthSign = myString.substring(2, 5);
        System.out.println(thirdToFifthSign);
    }

    public static void CheckIfStringContainsSubstring()
    {
        String myString = "LgbsKrkMnl";
        String mySubstring = "Krk";

        if (myString.contains(mySubstring))
            System.out.println("Contains");
        else
            System.out.println("Does not contain");
    }

    public static void GetPositionWhereSubsubstringAppearsInString()
    {
        String myString = "LgbsKrkMnl";
        String mySubstring = "Krk";

        int pos = myString.indexOf(mySubstring);

        System.out.println(pos);
    }

    public static void GetPositionWhereSubsubstringAppearsInString_FromRight()
    {
        String myString = "LgbsKrkMnlKrk";
        String mySubstring = "Krk";

        int pos = myString.lastIndexOf(mySubstring);

        System.out.println(pos);
    }

    public static void RestOfStringAfterKeyWord()
    {
        String myString = "LgbsKrkMnl";
        String mySubstring = "Krk";
        String restOfString = "";

        int firstSign = myString.indexOf(mySubstring);

        if (firstSign == -1)
            firstSign = 0;
        else
            firstSign = firstSign + 3;

        restOfString = myString.substring(firstSign);

        System.out.println(restOfString);
    }

    public static void ReplaceStringWithSubstring()
    {
        String myString = "LgbsKrkMnl";
        String replaceWhat = "Krk";
        String tobeReplacedWith = "hoho";

        myString = myString.replace(replaceWhat, tobeReplacedWith);

        System.out.println(myString);
    }

    public static void DateToStringWithFormat()
    {
        Date myDate = new Date();
        SimpleDateFormat format = new SimpleDateFormat("dd.MM.yyyy HH:mm:ss");
        String myString = format.format(myDate);
        System.out.println(myString);
    }

    public static void NumberToStringWithLeadingZeros()
    {
        int myInt = 33;
        int myInt2 = 3;
        String myString = "";
        
        myString = "000" + Integer.toString(myInt);
        myString = myString.substring(myString.length() - 3);
        System.out.println(myString);

        myString = "000" + Integer.toString(myInt2);
        myString = myString.substring(myString.length() - 3);
        System.out.println(myString);        
    }

    public static void CreateStringWithEnterAndCarriageReturn()
    {
        String myString = "abc\r\ndef\r\nghi";
        System.out.println(myString);

        myString = "abc" + (char)13 + (char)10 + "def" + (char)13 + (char)10 + "ghi";
        System.out.println(myString);
    }
}

File: 000_C# Training/C#_7 StringOperations/C#/MainClass.cs
using System;

class MainClass
{
    public static void Main()
    {
        CompareStrings_Convenient();
        CompareStrings_Correct();
        CompareString_CaseInsensitive();
        ConcatenateStrings();
        ConcatenateStringIteratively();
        LengthOfString();
        SplitStringBySingleChar();
        SplitStringByString();
        GetFirstFourSigns();
        GetFirstFourSigns_Safe();
        GetLastFourSigns();
        GetLastFourSigns_Safe();
        GetSignsBetween();
        CheckIfStringContainsSubstring();
        GetPositionWhereSubsubstringAppearsInString();
        GetPositionWhereSubsubstringAppearsInString_FromRight();
        RestOfStringAfterKeyWord();
        ReplaceStringWithSubstring();
        DateToStringWithFormat();
        NumberToStringWithLeadingZeros();
        CreateStringWithEnterAndCarriageReturn();

        Console.ReadLine();
    }

    public static void CompareStrings_Convenient()
    {
        string myString = "abc";
        string myString2 = "aef";

        if (myString == myString2)
            Console.WriteLine("Strings are equal");
        else
            Console.WriteLine("Strings are not equal");

        //Attention, in Java and C++, this approach would not work
        //See CompareStrings_Correct()
    }

    public static void CompareStrings_Correct()
    {
        string myString = "abc";
        string myString2 = "aef";

        if (myString.Equals(myString2))
            Console.WriteLine("Strings are equal");
        else
            Console.WriteLine("Strings are not equal");
    }        

    public static void CompareString_CaseInsensitive()
    {
        string myString = "AbCDeF";
        string myString2 = "abcdef";

        if (myString.ToLower() == myString2.ToLower())
            Console.WriteLine("Strings are equal");
        else
            Console.WriteLine("Strings are not equal");

        if (myString.ToUpper() == myString2.ToUpper())
            Console.WriteLine("Strings are equal");
        else
            Console.WriteLine("Strings are not equal");        

    }

    public static void ConcatenateStrings()
    {
        string myString = "Lgbs";
        string myString2 = "Mnl";
        string myString3 = "Krk";

        string myString4 = "We work at " + myString + " " + myString2 + " and " + myString3;
        
        Console.WriteLine(myString4);
    }

    public static void ConcatenateStringIteratively()
    {
        string[] myArray = new string[] {"abc", "def", "ghi", "jkl"};
        string myFinalString = "";

        foreach (string singleString in myArray)
            myFinalString += singleString + "__";  

        Console.WriteLine(myFinalString);
    }

    public static void LengthOfString()
    {
        string myString = "LgbsMnlKrk";
        int length = myString.Length;
        Console.WriteLine(length);
    }

    public static void SplitStringBySingleChar()
    {
        string myString = "abc;def;ghi;jkl";

        string[] mySplitString = myString.Split(';');

        foreach (string singleString in mySplitString)
            Console.WriteLine(singleString);
    }

    public static void SplitStringByString()
    {
        string myString = "abc;_def;_ghi;_jkl";

        string[] mySplitString = myString.Split(new string[] {";_"}, StringSplitOptions.None);

        foreach (string singleString in mySplitString)
            Console.WriteLine(singleString);        
    }

    public static void GetFirstFourSigns()
    {
        string myString = "LgbsKrkMnl";
        string leftFourSigns = myString.Substring(0, 4);
        Console.WriteLine(leftFourSigns);
        // Attention, when myString has less than four signs, an error will be thrown. 
        // See GetFirstFourSigns_Safe() to handle this problem
    }

    public static void GetFirstFourSigns_Safe()
    {
        string myString = "LgbsKrkMnl";
        string leftFourSigns = "";

        if (myString.Length >= 4)
            leftFourSigns = myString.Substring(0, 4);
        else
            leftFourSigns = myString;

        Console.WriteLine(leftFourSigns);
    }

    public static void GetLastFourSigns()
    {
        string myString = "LgbsKrkMnl";
        string rightFourSigns = myString.Substring(myString.Length - 4);
        Console.WriteLine(rightFourSigns);
    }

    public static void GetLastFourSigns_Safe()
    {
        string myString = "LgbsKrkMnl";
        string rightFourSigns = "";
    
        if (myString.Length >= 4)
            rightFourSigns = myString.Substring(myString.Length - 4);
        else
            rightFourSigns = myString;

        Console.WriteLine(rightFourSigns);
    }

    public static void GetSignsBetween()
    {
        string myString = "LgbsKrkMnl";
        string thirdToFifthSign = myString.Substring(2, 3);
        Console.WriteLine(thirdToFifthSign);
    }

    public static void CheckIfStringContainsSubstring()
    {
        string myString = "LgbsKrkMnl";
        string mySubstring = "Krk";

        if (myString.Contains(mySubstring))
            Console.WriteLine("Contains");
        else
            Console.WriteLine("Does not contain");
    }

    public static void GetPositionWhereSubsubstringAppearsInString()
    {
        string myString = "LgbsKrkMnl";
        string mySubstring = "Krk";

        int pos = myString.IndexOf(mySubstring);

        Console.WriteLine(pos);
    }

    public static void GetPositionWhereSubsubstringAppearsInString_FromRight()
    {
        string myString = "LgbsKrkMnlKrk";
        string mySubstring = "Krk";

        int pos = myString.LastIndexOf(mySubstring);

        Console.WriteLine(pos);
    }

    public static void RestOfStringAfterKeyWord()
    {
        string myString = "LgbsKrkMnl";
        string mySubstring = "Krk";
        string restOfString = "";

        int firstSign = myString.IndexOf(mySubstring);

        if (firstSign == -1)
            firstSign = 0;
        else
            firstSign = firstSign + 3;

        restOfString = myString.Substring(firstSign);

        Console.WriteLine(restOfString);
    }

    public static void ReplaceStringWithSubstring()
    {
        string myString = "LgbsKrkMnl";
        string replaceWhat = "Krk";
        string tobeReplacedWith = "hoho";

        myString = myString.Replace(replaceWhat, tobeReplacedWith);

        Console.WriteLine(myString);
    }

    public static void DateToStringWithFormat()
    {
        DateTime myDate = DateTime.Now;
        string myString = myDate.ToString("dd.MM.yyyy HH:mm:ss");
        Console.WriteLine(myString);
    }

    public static void NumberToStringWithLeadingZeros()
    {
        int myInt = 33;
        int myInt2 = 3;
        string myString = "";
        
        myString = "000" + myInt.ToString();
        myString = myString.Substring(myString.Length - 3);
        Console.WriteLine(myString);

        myString = "000" + myInt2.ToString();
        myString = myString.Substring(myString.Length - 3);
        Console.WriteLine(myString);        
    }

    public static void CreateStringWithEnterAndCarriageReturn()
    {
        string myString = "abc\r\ndef\r\nghi";
        Console.WriteLine(myString);

        myString = "abc" + (char)13 + (char)10 + "def" + (char)13 + (char)10 + "ghi";
        Console.WriteLine(myString);
    }
}